quinn-h3/src/proto/connection.rs

```diff
@@ -189,34 +195,54 @@ impl Connection {
         self.pending_streams[ty as usize].reserve(capacity);
     }
 
-    pub fn request_initiated(&mut self, id: StreamId) {
-        if !self.go_away {
-            self.requests_in_flight.insert(id);
-            self.max_id_in_flight = cmp::max(id.0, self.max_id_in_flight);
+    // The remote peer initiated a stream: a request from client or a push from server
+    pub fn remote_stream_initiated(&mut self, id: StreamId) -> Result<()> {
+        if let Some(max_id) = self.go_away {
+            if id.index() > max_id.index() {
+                warn!(id=%id, "rejecting");
+                return Err(Error::Aborted);
+            }
         }
+        debug!(id=%id, "accepting");
+        self.requests_in_flight.insert(id);
+        self.max_id_in_flight = if id.index() > self.max_id_in_flight.index() {
+            id
+        } else {
+            self.max_id_in_flight
+        };
+        Ok(())
     }
 
-    pub fn request_finished(&mut self, id: StreamId) {
-        if !self.go_away {
-            self.requests_in_flight.remove(&id);
-        }
+    // A stream initiated by the remote peer has finished
+    pub fn remote_stream_finished(&mut self, id: StreamId) {
+        trace!(id=%id, "finished");
+        self.requests_in_flight.remove(&id);
     }
 
-    pub fn requests_in_flight(&self) -> usize {
-        self.requests_in_flight.len()
-    }
-
-    pub fn go_away(&mut self) {
-        if !self.go_away {
-            self.go_away = true;
-            HttpFrame::Goaway(self.max_id_in_flight)
-                .encode(&mut self.pending_streams[PendingStreamType::Control as usize]);
-        }
+    // A graceful shutdown initiated locally, let `allow_streams` potentially in-flight
+    // incoming streams be processed.
+    pub fn go_away(&mut self, allow_streams: u64) {
+        let dir = if self.side == Side::Server {
```

Comment:
This is probably better written as a match (for one thing, it takes one less line).